File: scala/src/main/scala/fundamentals/Arrays.scala
package fundamentals

import scala.collection.mutable
import scala.collection.immutable.SortedSet

object Arrays:
  def demonstrate(): Unit =
    println("=== Arrays and Collections ===\n")
    
    // Arrays (mutable, fixed size)
    println("1. Arrays:")
    val numbers = Array(1, 2, 3, 4, 5)
    println(s"Original array: ${numbers.mkString(", ")}")
    numbers(0) = 10  // Modifying an element
    println(s"After modification: ${numbers.mkString(", ")}")
    
    // Array operations
    println("\nArray operations:")
    println(s"First element: ${numbers(0)}")
    println(s"Array length: ${numbers.length}")
    println(s"Sum of elements: ${numbers.sum}")
    println(s"Doubled elements: ${numbers.map(_ * 2).mkString(", ")}")
    
    // 2D Array
    println("\n2. 2D Array:")
    val matrix = Array.ofDim[Int](2, 3)
    for
      i <- matrix.indices
      j <- matrix(0).indices
    do
      matrix(i)(j) = i + j
    
    println("Matrix:")
    matrix.foreach(row => println(row.mkString(" ")))
    
    // Lists (immutable)
    println("\n3. Lists:")
    val fruits = List("apple", "banana", "orange")
    println(s"Original list: $fruits")
    
    // List operations
    println("\nList operations:")
    println(s"Head: ${fruits.head}")
    println(s"Tail: ${fruits.tail}")
    println(s"Is empty? ${fruits.isEmpty}")
    val moreFruits = "grape" :: fruits  // Prepending
    println(s"After prepending: $moreFruits")
    val combinedFruits = fruits ::: List("mango", "kiwi")  // Concatenation
    println(s"After concatenation: $combinedFruits")
    
    // Vector (immutable, good for random access)
    println("\n4. Vector:")
    val vector = Vector(1, 2, 3, 4, 5)
    println(s"Vector: $vector")
    println(s"Element at index 2: ${vector(2)}")
    val updatedVector = vector.updated(0, 10)
    println(s"After updating first element: $updatedVector")
    
    // Set (immutable by default)
    println("\n5. Sets:")
    val set1 = Set(1, 2, 3, 3, 4, 5, 5)  // Duplicates are removed
    println(s"Set1: $set1")
    val set2 = Set(4, 5, 6, 7)
    println(s"Set2: $set2")
    
    // Set operations
    println("\nSet operations:")
    println(s"Union: ${set1 union set2}")
    println(s"Intersection: ${set1 intersect set2}")
    println(s"Difference: ${set1 diff set2}")
    
    // SortedSet
    println("\n6. SortedSet:")
    val sortedSet = SortedSet(5, 2, 8, 1, 9, 3)
    println(s"SortedSet: $sortedSet")
    
    // Map (immutable by default)
    println("\n7. Maps:")
    val capitals = Map(
      "USA" -> "Washington DC",
      "UK" -> "London",
      "France" -> "Paris"
    )
    println(s"Capitals: $capitals")
    
    // Map operations
    println("\nMap operations:")
    println(s"Capital of France: ${capitals("France")}")
    println(s"Contains UK? ${capitals.contains("UK")}")
    println(s"All keys: ${capitals.keys}")
    println(s"All values: ${capitals.values}")
    
    // Mutable collections
    println("\n8. Mutable Collections:")
    val mutableList = mutable.ListBuffer(1, 2, 3)
    println(s"Original mutable list: $mutableList")
    mutableList += 4  // Adding element
    mutableList -= 2  // Removing element
    println(s"After modifications: $mutableList")
    
    // Mutable Set
    val mutableSet = mutable.Set(1, 2, 3)
    println(s"\nMutable set: $mutableSet")
    mutableSet += 4  // Adding element
    mutableSet -= 2  // Removing element
    println(s"After modifications: $mutableSet")
    
    // Mutable Map
    val mutableMap = mutable.Map("a" -> 1, "b" -> 2)
    println(s"\nMutable map: $mutableMap")
    mutableMap("c") = 3  // Adding new key-value pair
    mutableMap -= "a"    // Removing key-value pair
    println(s"After modifications: $mutableMap")
    
    // Sequences
    println("\n9. Sequences:")
    val seq = Seq(1, 2, 3, 4, 5)
    println(s"Sequence: $seq")
    println(s"First element: ${seq.head}")
    println(s"Last element: ${seq.last}")
    println(s"Length: ${seq.length}")
    
    // Range
    println("\n10. Range:")
    val range1 = 1 to 5
    println(s"Range (inclusive): ${range1.mkString(", ")}")
    val range2 = 1 until 5
    println(s"Range (exclusive): ${range2.mkString(", ")}")
    val range3 = 1 to 10 by 2
    println(s"Range with step: ${range3.mkString(", ")}")
    
    // Collection operations
    println("\n11. Common Collection Operations:")
    val numbers2 = List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
    println("Original list: " + numbers2)
    println(s"Mapped (x * 2): ${numbers2.map(_ * 2)}")
    println(s"Filtered (even): ${numbers2.filter(_ % 2 == 0)}")
    println(s"Reduced (sum): ${numbers2.reduce(_ + _)}")
    println(s"Folded (sum): ${numbers2.fold(0)(_ + _)}")
    println(s"Grouped by even/odd: ${numbers2.groupBy(_ % 2 == 0)}")
    
    println() // Empty line

File: kotlin/src/main/kotlin/fundamentals/Arrays.kt
package fundamentals

class Arrays {
    fun demonstrate() {
        println("=== Arrays and Collections ===\n")

        // Arrays
        println("1. Arrays:")
        val numbers = Array(5) { it * 2 }  // [0, 2, 4, 6, 8]
        println("Numbers array: ${numbers.contentToString()}")

        // Array of primitive types
        val intArray = IntArray(5) { it + 1 }  // [1, 2, 3, 4, 5]
        val doubleArray = DoubleArray(3) { it * 1.5 }  // [0.0, 1.5, 3.0]
        println("Int array: ${intArray.contentToString()}")
        println("Double array: ${doubleArray.contentToString()}")

        // Array operations
        println("\n2. Array operations:")
        println("First element: ${numbers[0]}")
        println("Last element: ${numbers[numbers.size - 1]}")
        println("Size: ${numbers.size}")
        numbers[0] = 10
        println("After setting first element to 10: ${numbers.contentToString()}")

        // Lists
        println("\n3. Lists:")
        val readOnlyList = listOf(1, 2, 3, 4, 5)
        val mutableList = mutableListOf(1, 2, 3)
        println("Read-only list: $readOnlyList")
        println("Mutable list: $mutableList")

        // List operations
        mutableList.add(4)
        mutableList.removeAt(0)
        println("After adding 4 and removing first element: $mutableList")
        println("First element: ${mutableList.first()}")
        println("Last element: ${mutableList.last()}")
        println("Contains 3? ${mutableList.contains(3)}")

        // Sets
        println("\n4. Sets:")
        val readOnlySet = setOf(1, 2, 2, 3, 3, 3)  // Duplicates are removed
        val mutableSet = mutableSetOf(1, 2, 3)
        println("Read-only set: $readOnlySet")
        println("Mutable set: $mutableSet")

        // Set operations
        mutableSet.add(4)
        mutableSet.remove(1)
        println("After adding 4 and removing 1: $mutableSet")
        println("Contains 2? ${mutableSet.contains(2)}")

        // Set mathematical operations
        val set1 = setOf(1, 2, 3, 4)
        val set2 = setOf(3, 4, 5, 6)
        println("\n5. Set operations:")
        println("Set1: $set1")
        println("Set2: $set2")
        println("Union: ${set1.union(set2)}")
        println("Intersection: ${set1.intersect(set2)}")
        println("Difference (set1 - set2): ${set1.subtract(set2)}")

        // Maps
        println("\n6. Maps:")
        val readOnlyMap = mapOf(
            "one" to 1,
            "two" to 2,
            "three" to 3
        )
        val mutableMap = mutableMapOf(
            "apple" to 1,
            "banana" to 2,
            "orange" to 3
        )
        println("Read-only map: $readOnlyMap")
        println("Mutable map: $mutableMap")

        // Map operations
        mutableMap["grape"] = 4
        mutableMap.remove("apple")
        println("After adding grape and removing apple: $mutableMap")
        println("Value for banana: ${mutableMap["banana"]}")
        println("Contains key 'orange'? ${mutableMap.containsKey("orange")}")
        println("Contains value 4? ${mutableMap.containsValue(4)}")

        // Map iteration
        println("\n7. Map iteration:")
        for ((key, value) in mutableMap) {
            println("$key = $value")
        }

        // List transformations
        println("\n8. List transformations:")
        val numbers2 = (1..5).toList()
        println("Original numbers: $numbers2")
        println("Doubled: ${numbers2.map { it * 2 }}")
        println("Even numbers: ${numbers2.filter { it % 2 == 0 }}")
        println("Sum: ${numbers2.sum()}")
        println("Product: ${numbers2.reduce { acc, num -> acc * num }}")

        // Collection grouping
        println("\n9. Collection grouping:")
        val words = listOf("apple", "banana", "apricot", "orange", "avocado", "blueberry")
        val grouped = words.groupBy { it.first() }
        println("Grouped by first letter: $grouped")

        // Collection flattening
        println("\n10. Collection flattening:")
        val nestedList = listOf(
            listOf(1, 2, 3),
            listOf(4, 5, 6),
            listOf(7, 8, 9)
        )
        println("Nested list: $nestedList")
        println("Flattened: ${nestedList.flatten()}")

        // Sequence operations
        println("\n11. Sequence operations:")
        val sequence = generateSequence(1) { it + 1 }
            .take(5)
            .filter { it % 2 == 0 }
            .map { it * it }
        println("Sequence result: ${sequence.toList()}")

        println() // Empty line
    }
}